lens/history.go

```diff
@@ -0,0 +1,279 @@
+// Copyright 2023 Democratized Data Foundation
+//
+// Use of this software is governed by the Business Source License
+// included in the file licenses/BSL.txt.
+//
+// As of the Change Date specified in that file, in accordance with
+// the Business Source License, use of this software will be governed
+// by the Apache License, Version 2.0, included in the file
+// licenses/APL.txt.
+
+package lens
+
+import (
+	"context"
+
+	"github.com/ipfs/go-datastore/query"
+	"github.com/sourcenetwork/immutable"
+
+	"github.com/sourcenetwork/defradb/client"
+	"github.com/sourcenetwork/defradb/core"
+	"github.com/sourcenetwork/defradb/datastore"
+)
+
+// schemaHistoryLink represents an item in a particular schema's history, it
+// links to the previous and next version items if they exist.
+type schemaHistoryLink struct {
+	// The schema version id of this history item.
+	schemaVersionID string
+
+	// The history link to the next schema version, if there is one
+	// (for the most recent schema version this will be None).
+	next immutable.Option[*schemaHistoryLink]
+
+	// The history link to the previous schema version, if there is
+	// one (for the initial schema version this will be None).
+	previous immutable.Option[*schemaHistoryLink]
+}
+
+// targetedSchemaHistoryLink represents an item in a particular schema's history, it
+// links to the previous and next version items if they exist.
+//
+// It also contains a vector which describes the distance and direction to the
+// target schema version (given as an input param on construction).
+type targetedSchemaHistoryLink struct {
+	// The schema version id of this history item.
+	schemaVersionID string
+
+	// The link to next schema version, if there is one
+	// (for the most recent schema version this will be None).
+	next immutable.Option[*targetedSchemaHistoryLink]
+
+	// The link to the previous schema version, if there is
+	// one (for the initial schema version this will be None).
+	previous immutable.Option[*targetedSchemaHistoryLink]
+
+	// The distance and direction from this history item to the target.
+	//
+	// A zero value indicates that this is the target item. A positive value
+	// indicates that the target is more recent. A negative value indicates
+	// that the target predates this history item.
+	targetVector int
```

Comment:
suggestion: wouldn't `distanceToTarget` fit more as a name?

Reply:
No. `distance` lacks direction, this property contains direction.

Reply:
Hmm, don't agree here, this is a vector as you noted, but its a 1-dimensional vector, which reduces down to a single positive/negative direction, I do agree its better. `Vector` sounds a little more heavy/loaded of a term, and that there is a slighty more involved usage. (Slight preference)

Reply:
A one-dimensional vector is called a scalar :sweat_smile:  Is also worth highlighting here that the direction is the only component of this vector that is actually used at the moment.  The distance component is really deadcode atm.

Reply:
> A one-dimensional vector is called a scalar

Yes and no, a vector is direction and magnitude, which can be expressed using 2 scalars (if you have a 2D space), but really you just need something to convey direction and magnitude. But, since this is a vector that only lies on a single dimension (1D space), the "direction" is simply the positive/negative sign, and the "magnitude" is the actual value.

Obviously this is getting a little into the weeds vector math and such, but this I think highlights the problem of calling this a vector, *especially* if as you say, only the direction is being used at the moment.